### PR TITLE
Report FQDN on host.name enabled by a feature flag

### LINKED ISSUE #1070: Report host name as FQDN on all platforms
Apparently Go does not provide the FQDN on all systems (particularly Linux, it seems) because it checks `/bin/hostname` rather than `/bin/hostname -f`.

Here's the Go test showing that they accept the shortname: https://golang.org/src/os/os_test.go#L1341

It would be great if we could fill in the actual FQDN rather than an unexpected subset, but it does mean providing an appropriate mechanism for _every_ supported system (or falling back to the subpar behavior that comes out of the box).

As a workaround, #715 will allow environment variables to fill this in, but it still requires something else in the environment to be used.

/cc @andrewkroh

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

This PR enhances Beats to accept feature flag configuration either in standalone or managed mode.  It also enhances Beats to report their host's FQDN when the FQDN feature flag is set.

For Beats running in standalone mode, feature flag configuration is defined in the `<beat>.yml` configuration file as follows:

```yaml
features:
  <flagname>:
    enabled: true
```

For example, to run Metricbeat standalone with the FQDN feature flag enabled, configure the following section in `metricbeat.yml`:

```yaml
features:
  fqdn:
    enabled:  true
```

## Why is it important?

To enable use cases, particularly involving running multiple Beats in a containerized environment, where it's useful for every Beat to report it's host's FQDN.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

#### With FQDN feature flag disabled

https://user-images.githubusercontent.com/51061/223835005-aceafafe-c3e1-4b61-883e-a58af2883b09.mov

#### With FQDN feature flag enabled

https://user-images.githubusercontent.com/51061/223835046-614d43ca-bca0-4949-a4d0-0ca5e98606ea.mov

#### With FQDN feature flag disabled

https://user-images.githubusercontent.com/51061/223835856-bd8d486d-b9f2-4746-af56-afdb69356ee8.mov

#### With FQDN feature flag enabled

https://user-images.githubusercontent.com/51061/223836401-e1496f2f-ca60-4e71-a859-deadf3458754.mov

## How to test this PR locally

See screen recordings in previous section.

## Related issues

- Closes #1070 

-

### CODE REVIEW

**Comment by cmacknz** [comment]
There are a few other references to the `host.name` field besides just the one in `add_host_metadata`. 

`rg '"host.name"' -g '!*_test.go' -g '!*.py' -g '!*.json'` will find them using [ripgrep](https://github.com/BurntSushi/ripgrep).

**Comment by cmacknz** [comment]
It appears that the Windows event logs might be the FQDN by default. We should explicitly try to test and inspect the `host.name` value for each path in Beats that sets it to ensure it is consistent everywhere.

**Comment by AndersonQ** [comment]
> There are a few other references to the `host.name` field besides just the one in `add_host_metadata`.
> 
> `rg '"host.name"' -g '!*_test.go' -g '!*.py' -g '!*.json'` will find them using [ripgrep](https://github.com/BurntSushi/ripgrep).
> 
> ```
> libbeat/processors/decode_xml_wineventlog/processor.go
> libbeat/processors/add_host_metadata/add_host_metadata.go
> libbeat/docs/troubleshooting/default_field.md
> x-pack/filebeat/module/imperva/securesphere/config/liblogparser.js
> x-pack/filebeat/module/fortinet/clientendpoint/config/liblogparser.js
> x-pack/filebeat/module/fortinet/fortimanager/config/liblogparser.js
> x-pack/filebeat/module/fortinet/fortimail/config/liblogparser.js
> x-pack/filebeat/module/snort/log/config/liblogparser.js
> x-pack/filebeat/module/infoblox/nios/config/liblogparser.js
> x-pack/filebeat/module/radware/defensepro/config/liblogparser.js
> x-pack/metricbeat/module/aws/cloudwatch/metadata/ec2/ec2.go
> x-pack/filebeat/module/sophos/utm/config/liblogparser.js
> x-pack/filebeat/module/juniper/junos/config/liblogparser.js
> x-pack/filebeat/module/juniper/netscreen/config/liblogparser.js
> x-pack/filebeat/module/zeek/dhcp/test/dhcp-json.log
> x-pack/filebeat/module/o365/audit/config/pipeline.js
> x-pack/filebeat/module/sonicwall/firewall/config/liblogparser.js
> x-pack/filebeat/module/bluecoat/director/config/liblogparser.js
> x-pack/filebeat/module/netscout/sightline/config/liblogparser.js
> x-pack/filebeat/module/barracuda/waf/config/liblogparser.js
> x-pack/filebeat/module/barracuda/spamfirewall/config/liblogparser.js
> x-pack/filebeat/module/cisco/nexus/config/liblogparser.js
> x-pack/filebeat/module/cisco/meraki/config/liblogparser.js
> x-pack/filebeat/module/f5/bigipafm/config/liblogparser.js
> x-pack/filebeat/module/proofpoint/emailsecurity/config/liblogparser.js
> x-pack/filebeat/module/f5/bigipapm/config/liblogparser.js
> x-pack/filebeat/module/microsoft/dhcp/config/liblogparser.js
> x-pack/filebeat/module/squid/log/config/liblogparser.js
> x-pack/filebeat/module/tomcat/log/config/liblogparser.js
> x-pack/metricbeat/module/gcp/metrics/metricset.go
> x-pack/filebeat/module/crowdstrike/falcon/ingest/firewall_match.yml
> x-pack/filebeat/module/zscaler/zia/config/liblogparser.js
> x-pack/metricbeat/module/azure/add_metadata.go
> x-pack/filebeat/module/cylance/protect/config/liblogparser.js
> winlogbeat/eventlog/eventlog.go
> metricbeat/module/beat/state/data.go
> ```

luckily the huge majority of these files are the same, `liblogparser.js`, and so far I don't think we'll need to modify them.

There is indeed a few places that indeed use a different approach to populate `host.name` and [winlogbeat](https://github.com/elastic/kibana/pull/150239#issuecomment-1423031598) is one of them.

**Comment by ycombinator** [comment]
Tested this PR with the `add_host_metadata` processor and it works as expected:

```
$ hostname
shaunak-test-fqdn

$ hostname -f
shaunak-test-fqdn.c.elastic-obs-integrations-dev.internal
```

```
$ cat filebeat.fqdn.disabled.yml
filebeat.inputs:
- type: stdin

processors:
- add_host_metadata: ~

output:
  console:
    enabled: true
```

```
$ echo "$RANDOM message" \
> | ./filebeat-linux-amd64 -c filebeat.fqdn.disabled.yml \
> | jq -c '{name: .host.name, hostname: .host.hostname, domain: .host.domain, message: .message}'
{"name":"shaunak-test-fqdn","hostname":"shaunak-test-fqdn","domain":null,"message":"1668 message"}
^C
```

```
$ cat filebeat.fqdn.enabled.yml
filebeat.inputs:
- type: stdin

processors:
- add_host_metadata: ~

features:
  fqdn:
    enabled: true

output:
  console:
    enabled: true
```

```
$ echo "$RANDOM message" \
> | ./filebeat-linux-amd64 -c filebeat.fqdn.enabled.yml \
> | jq -c '{name: .host.name, hostname: .host.hostname, domain: .host.domain, message: .message}'
{"name":"shaunak-test-fqdn.c.elastic-obs-integrations-dev.internal","hostname":"shaunak-test-fqdn","domain":null,"message":"2512 message"}
```

File: winlogbeat/sys/wineventlog/renderer.go [hunk 1]
```diff
@@ -222,6 +222,7 @@ func (r *Renderer) renderSystem(handle EvtHandle, event *winevent.Event) error {
 			event.Channel = data.(string)
 		case EvtSystemComputer:
 			event.Computer = data.(string)
+			// TODO: set event.Computer to FQDN if features.FQDN() is true?
```
Comment (ycombinator) [inline winlogbeat/sys/wineventlog/renderer.go:225]: @efd6 @andrewkroh In this PR we're trying to set `host.name` in events generated by Beats to the FQDN if the Beat config contains `features.fqdn.enabled: true`.  What's the best way to retrieve the FQDN on a Windows host? I assume this is the line in code where we'd want to set `event.Computer` to the FQDN if the feature is enabled?

**Comment by ycombinator** [comment]
@cmacknz Thanks for posting the `rg` command.  I re-ran it, but also excluded files under Beat module folders and `*.md` files, and here' what comes up:

```
$ rg '"host.name"' -g '!*_test.go' -g '!*.py' -g '!*.json' -g '!*.md' -g '!**/*beat/module/**'
winlogbeat/eventlog/eventlog.go
101:	winevent.AddOptional(m, "host.name", e.Computer)

libbeat/processors/decode_xml_wineventlog/processor.go
149:	winevent.AddOptional(ecs, "host.name", evt.Computer)

libbeat/processors/add_host_metadata/add_host_metadata.go
152:		data.Put("host.name", p.config.Name)
```

For the first two results, related to hostnames on Windows, I'll follow up in  https://github.com/elastic/beats/pull/34456#discussion_r1127243121.

That leaves the last result, related to the `add_host_metadata` processor.  This area has already been addressed in this PR.

Additionally, not seen in the `rg` results, there is also the hostname reported by the `GET /` and `GET /state` Beats HTTP APIs, which are called by the `beat` Metricbeat module, which Agent uses for monitoring it's component Beats' health.  This area has also been addressed in this PR.

File: winlogbeat/sys/wineventlog/renderer.go (hunk 1 above)
Comment (ycombinator) [reply winlogbeat/sys/wineventlog/renderer.go:225]: Thanks for the background and suggestions, @andrewkroh.  I've created a separate issue to handle FQDN reporting for Windows hosts, as the use case urgently driving the FQDN feature implementation is for Beats running in containerized environments, which tend to be *nix most of the times.  I've created https://github.com/elastic/beats/issues/34782 to follow up on implementing FQDN reporting for Windows hosts.

**Comment by michalpristas** [comment]
there are some windows build failures please fix those

File: dev-tools/notice/overrides.json [hunk 2]
```diff
@@ -1,4 +1,6 @@
 {"name": "github.com/elastic/elastic-agent-client/v7", "licenceType": "Elastic"}
+{"name": "github.com/AndersonQ/elastic-agent-client/v7", "licenceType": "Elastic"}
```
Comment (michalpristas) [inline dev-tools/notice/overrides.json:2]: are you referencing own repos?

File: NOTICE.txt [hunk 3]
```diff
@@ -9866,12 +9866,12 @@ Contents of probable licence file $GOMODCACHE/github.com/elastic/elastic-agent-a
 
 
 --------------------------------------------------------------------------------
-Dependency : github.com/elastic/elastic-agent-client/v7
-Version: v7.0.3
+Dependency : github.com/AndersonQ/elastic-agent-client/v7
```
Comment (michalpristas) [inline NOTICE.txt:9869]: is this on purpose?

File: go.mod [hunk 4]
```diff
@@ -377,3 +376,7 @@ replace (
 
 // Exclude this version because the version has an invalid checksum.
 exclude github.com/docker/distribution v2.8.0+incompatible
+
+replace github.com/elastic/elastic-agent-client/v7 => github.com/AndersonQ/elastic-agent-client/v7 v7.0.0-20230308191952-d7d45d93d0da
```
Comment (michalpristas) [inline go.mod:380]: i dont' really like this

File: libbeat/cmd/instance/beat.go [hunk 5]
```diff
@@ -242,6 +244,12 @@ func NewBeat(name, indexPrefix, v string, elasticLicensed bool) (*Beat, error) {
 		return nil, err
 	}
 
+	h, err := sysinfo.Host()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get FQDN: %w", err)
```
Comment (michalpristas) [inline libbeat/cmd/instance/beat.go:249]: aren't we getting fqdn in a next step?

File: libbeat/features/features.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+// UpdateFromProto updates the feature flags configuration. If f is nil UpdateFromProto is no-op.
+func UpdateFromProto(f *proto.Features) {
+	if f == nil {
+		logp.L().Debug("feature flags are nil, ignoring them")
+		return
+	}
+
+	if f.Fqdn == nil {
+		f.Fqdn = &proto.FQDNFeature{}
```
Comment (michalpristas) [inline libbeat/features/features.go:47]: could we make it more explicit it is disabled by default?

File: libbeat/features/features.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	parsedFlags := cfg{}
+	if err := c.Unpack(&parsedFlags); err != nil {
+		return fmt.Errorf("could not Unpack features config: %w", err)
+	}
+
+	flags.mu.Lock()
+	defer flags.mu.Unlock()
+	flags.fqdnEnabled = parsedFlags.Features.FQDN.Enabled()
```
Comment (michalpristas) [inline libbeat/features/features.go:75]: can we hit nil ref with FQDN or features?

File: libbeat/features/features_test.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+			name: "FQDN empty",
+			yaml: `
+  features:
+    fqdn:`,
+			want: false,
+		},
+		{
+			name: "FQDN absent",
+			yaml: `
+  features:`,
```
Comment (michalpristas) [inline libbeat/features/features_test.go:65]: please test config without features

File: libbeat/publisher/processing/default.go [hunk 9]
```diff
@@ -143,10 +144,17 @@ var WithHost modifier = builtinModifier(func(info beat.Info) mapstr.M {
 // pipeline.
 func WithAgentMeta() modifier {
 	return builtinModifier(func(info beat.Info) mapstr.M {
+		var hostname string
+		if features.FQDN() {
```
Comment (michalpristas) [inline libbeat/publisher/processing/default.go:148]: small nit: this check is on different places, maybe we could provide a helper in info accepting a flag and returning proper hostname

File: go.mod (hunk 4 above)
Comment (michalpristas) [reply go.mod:380]: i see it's not in yet, doing PR reviews in reverse. my bad

File: dev-tools/notice/overrides.json (hunk 2 above)
Comment (ycombinator) [reply dev-tools/notice/overrides.json:2]: Yes, only until https://github.com/elastic/elastic-agent-client/pull/56 is merged.  After that we can remove this line.

File: go.mod (hunk 4 above)
Comment (ycombinator) [reply go.mod:380]: No worries.  Indeed, once https://github.com/elastic/elastic-agent-client/pull/56 is merged, we can remove this line.

File: dev-tools/notice/overrides.json (hunk 2 above)
Comment (ycombinator) [reply dev-tools/notice/overrides.json]: Removed.

File: go.mod (hunk 4 above)
Comment (ycombinator) [reply go.mod]: Removed.

File: x-pack/libbeat/management/managerV2.go [hunk 10]
```diff
@@ -560,16 +582,17 @@ func (cm *BeatV2Manager) reloadOutput(unit *client.Unit) error {
 		return nil
 	}
 
-	reloadConfig, err := groupByOutputs(rawConfig)
+	reloadConfig, err := groupByOutputs(expected.Config)
 	if err != nil {
 		return fmt.Errorf("failed to generate config for output: %w", err)
 	}
 
+	// features.UpdateFromProto(featureFlags)
```
Comment (cmacknz) [inline x-pack/libbeat/management/managerV2.go:590]: Should this still be here?

**Review by cmacknz — COMMENTED** [review]
I think we need some more tests of the changes to the libbeat/management code. There are some tests that pass sample unit expected configurations through it in https://github.com/elastic/beats/blob/874ba7b6af471ae2ad017e7a202b2bbab2229f79/x-pack/libbeat/management/managerV2_test.go#L26 that need to account for the new features behavior.

Another good way to test this code is to enroll an agent using this version of Beats into Fleet and start changing things in the UI and ensure everything keeps working:
- Updating the agent log level.
- Adding and removing integrations.
- Editing integrations.
- Enabling and disabling the feature flags.

File: go.mod [hunk 11]
```diff
@@ -60,17 +60,17 @@ require (
...
-	github.com/docker/go-units v0.4.0
+	github.com/docker/go-units v0.5.0
 	github.com/dolmen-go/contextio v0.0.0-20200217195037-68fc5150bcd5
 	github.com/dop251/goja v0.0.0-20200831102558-9af81ddcf0e1
 	github.com/dop251/goja_nodejs v0.0.0-20171011081505-adff31b136e6
 	github.com/dustin/go-humanize v1.0.0
 	github.com/eapache/go-resiliency v1.2.0
 	github.com/eclipse/paho.mqtt.golang v1.3.5
-	github.com/elastic/elastic-agent-client/v7 v7.0.3
+	github.com/elastic/elastic-agent-client/v7 v7.0.3-0.20230309151742-db7f3a085cc0
```
[the remark this answers is not in the document]
Comment (ycombinator) [reply go.mod:73]: Yes, I am planning to do this after this PR and https://github.com/elastic/elastic-agent/pull/2218 are merged, in case we need more changes to github.com/elastic/elastic-agent-client en route to the merges.

File: libbeat/features/features.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+type fflags struct {
+	mu sync.RWMutex
+
+	fqdnEnabled bool
+}
+
+// UpdateFromProto updates the feature flags configuration. If f is nil UpdateFromProto is no-op.
+func UpdateFromProto(f *proto.Features) {
+	if f == nil {
+		logp.L().Debug("feature flags are nil, ignoring them")
```
[the remark this answers is not in the document]
Comment (ycombinator) [reply libbeat/features/features.go]: Removed this line and a similar line from the `UpdateFromConfig` function in the same file.

File: x-pack/libbeat/management/managerV2.go (hunk 10 above)
Comment (ycombinator) [reply x-pack/libbeat/management/managerV2.go:590]: No, removed.

File: x-pack/libbeat/management/managerV2.go [hunk 13]
```diff
@@ -443,24 +459,30 @@ func (cm *BeatV2Manager) reload(units map[unitKey]*client.Unit) {
...
+		if expected.LogLevel > lowestLevel {
 			// log level is still used from an expected stopped unit until
 			// the unit is completely removed (aka. fully stopped)
-			lowestLevel = ll
+			lowestLevel = expected.LogLevel
 		}
-		if state == client.UnitStateStopped {
+		if expected.Features != nil {
+			// unit is expected to update its feature flags
+			features.UpdateFromProto(expected.Features)
```
[the remark this answers is not in the document]
Comment (ycombinator) [reply x-pack/libbeat/management/managerV2.go:472]: I've tested this manually via Fleet and it works as expected.  I created a policy where the FQDN feature flag was enabled, assigned it to the enrolled Agent, and set the Agent's log level to `debug`.  Once debug-level logs started streaming in, I updated the Agent's policy to disable the FQDN feature flag.  I verified that debug-level logs continued to stream in but now they contained the short hostname instead.  See screenshots below.

Screenshot 2023-03-09 at 13 19 09

Screenshot 2023-03-09 at 13 18 46
Screenshot 2023-03-09 at 13 18 52

**Comment by ycombinator** [comment]
> Another good way to test this code is to enroll an agent using this version of Beats into Fleet and start changing things in the UI and ensure everything keeps working:

> Updating the agent log level.
> Adding and removing integrations.
> Editing integrations.
> Enabling and disabling the feature flags.

Yes, I've manually tested Agent built with this PR and its associates. enrolled it in Fleet, and made changes to a) FQDN feature flag and b) monitoring one change at a time as well as multiple changes at a time.  In all cases, Agent behaved as expected. 

I did notice with toggling the FQDN feature flag back and forth that sometimes this change would take a few minutes to take effect, i.e. all the documents in `logs-*` and `metrics-*` to start reporting `host.name` appropriately.  At other times, though, the change took effect within a few seconds.  But in every case (I manually tested 11 policy changes), the changes did eventually take effect as expected.

File: libbeat/features/features.go (hunk 6 above)
Comment (ycombinator) [reply libbeat/features/features.go]: Done.

File: libbeat/features/features.go (hunk 7 above)
Comment (ycombinator) [reply libbeat/features/features.go:75]: We can't hit a nil ref with `parsedFlags.Features` because it's a struct (not pointer).  But good catch on `parsedFlags.Features.FQDN`.  I will add a guard.  Thanks.

Comment (ycombinator) [reply libbeat/features/features.go:75]: Actually, if `parsedFlags.Features.FQDN` is `nil`, `parsedFlags.Features.FQDN.Enabled()` will still handle it and return `false`:

https://github.com/elastic/elastic-agent-libs/blob/0ba2aa0ce1ac3a7a3dd511854e1cc65463784bb4/config/config.go#L246-L248

So I think we're good here? Or do you think we should add a guard anyway?

File: libbeat/features/features_test.go (hunk 8 above)
Comment (ycombinator) [reply libbeat/features/features_test.go:65]: Added a case for this.

File: libbeat/publisher/processing/default.go (hunk 9 above)
Comment (ycombinator) [reply libbeat/publisher/processing/default.go]: Good idea, done.

File: libbeat/processors/add_host_metadata/add_host_metadata.go [hunk 14]
```diff
@@ -134,16 +135,23 @@ func (p *addHostMetadata) loadData() error {
...
+				return fmt.Errorf("could not set host.mac: %w", err)
+			}
 		}
 	}
 
 	if p.config.Name != "" {
-		data.Put("host.name", p.config.Name)
+		if _, err := data.Put("host.name", p.config.Name); err != nil {
+			return fmt.Errorf("could not set host.name: %w", err)
+		}
```
Comment (ycombinator) [inline libbeat/processors/add_host_metadata/add_host_metadata.go:152]: This change is unrelated to this PR.  It's been made to please the linter.

File: libbeat/processors/add_host_metadata/add_host_metadata.go [hunk 15]
```diff
@@ -57,20 +56,22 @@ const (
 
 // New constructs a new add_host_metadata processor.
 func New(cfg *config.C) (processors.Processor, error) {
-	config := defaultConfig()
-	if err := cfg.Unpack(&config); err != nil {
-		return nil, errors.Wrapf(err, "fail to unpack the %v configuration", processorName)
+	c := defaultConfig()
+	if err := cfg.Unpack(&c); err != nil {
+		return nil, fmt.Errorf("fail to unpack the %v configuration: %w", processorName, err)
```
Comment (ycombinator) [inline libbeat/processors/add_host_metadata/add_host_metadata.go:61]: The change on this line is unrelated to this PR.  It's been made to please the linter.

File: libbeat/processors/add_host_metadata/add_host_metadata.go [hunk 16]
```diff
@@ -22,9 +22,8 @@ import (
 	"sync"
 	"time"
 
-	"github.com/pkg/errors"
-
```
Comment (ycombinator) [inline libbeat/processors/add_host_metadata/add_host_metadata.go:26]: This change is unrelated to this PR.  It's been made to please the linter.

File: x-pack/libbeat/management/managerV2.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (cmacknz) [reply x-pack/libbeat/management/managerV2.go:472]: Great, it looked right in the code. Thanks for checking.

**Comment by cmacknz** [comment]
> I did notice with toggling the FQDN feature flag back and forth that sometimes this change would take a few minutes to take effect, i.e. all the documents in logs-* and metrics-* to start reporting host.name appropriately. At other times, though, the change took effect within a few seconds. But in every case (I manually tested 11 policy changes), the changes did eventually take effect as expected.

add_host_metadata has a basic caching mechanism that is probably causing this: https://github.com/elastic/beats/blob/6e0346568e840eaa722b7174a1a073e520058cb3/libbeat/processors/add_host_metadata/add_host_metadata.go#L118-L121

It might be worth setting the cache to expire immediately when the feature flag changes. This will prevent confusion from users who try this themselves. If we can't change this, then we should document that the delay is expected (assuming the caching here is causing it).

**Comment by ycombinator** [comment]
> add_host_metadata has a basic caching mechanism that is probably causing [the delay]...

Thanks for this pointer, @cmacknz.  Let me try your suggestion of immediately expiring the cache upon FQDN feature flag change and see if that has any effect in the speed at which the impact of the feature flag change happens.

**Comment by ycombinator** [comment]
@cmacknz I temporarily disabled the cache in the `add_host_metadata` processor just to test your theory and that's totally what's going on here.  I switched between `fqdn-enabled` and `fqdn-disabled` policies a few times within the space of 5 minutes (which, BTW, is the default TTL for the cache in question), and you can see that the changes in the reported hostname took effect immediately:

Screenshot 2023-03-09 at 16 34 15

I'll add some code to this PR to expire the cache when the FQDN feature flag value changes.

**Comment by ycombinator** [comment]
@cmacknz I implemented the `add_host_metadata` cache expiry upon FQDN feature flag value change in 7dc7b49d77.  Please take a look and let me know what you think.  Thanks!

File: libbeat/features/features.go [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+	type cfg struct {
+		Features struct {
+			FQDN *conf.C `json:"fqdn" yaml:"fqdn" config:"fqdn"`
+		} `json:"features" yaml:"features" config:"features"`
+	}
+
+	parsedFlags := cfg{}
+	if err := c.Unpack(&parsedFlags); err != nil {
+		return fmt.Errorf("could not Unpack features config: %w", err)
```
Comment (michalpristas) [inline libbeat/features/features.go:73]: minor: s/Unpack/unpack

**Review by michalpristas — APPROVED** [review]
seems ok now

File: libbeat/features/features.go (hunk 17 above)
Comment (ycombinator) [reply libbeat/features/features.go]: Fixed.

**Comment by michalpristas** [comment]
seems like add host metadata failures can be related in integration tests

**Comment by ycombinator** [comment]
> seems like add host metadata failures can be related in integration tests

Yes, I'm able to reproduce this failure locally.  Working through it now...

**Comment by dedemorton** [comment]
I will fix the doc build errors and push the commit when I'm done.

File: libbeat/processors/add_host_metadata/add_host_metadata_test.go [hunk 18]
```diff
@@ -467,3 +469,40 @@ func TestSkipAddingHostMetadata(t *testing.T) {
 		})
 	}
 }
+
+func TestExpireCacheOnFQDNReportingChange(t *testing.T) {
```
Comment (cmacknz) [inline libbeat/processors/add_host_metadata/add_host_metadata_test.go:473]: 👍

File: winlogbeat/sys/wineventlog/renderer.go (hunk 1 above)
Comment (cmacknz) [reply winlogbeat/sys/wineventlog/renderer.go:225]: Do we need to keep this TODO in the code? From our previous discussion I believe we concluded we likely don't need to touch this.

File: x-pack/libbeat/management/managerV2.go [hunk 19]
```diff
@@ -385,8 +398,12 @@ func (cm *BeatV2Manager) unitListen() {
 			cm.UpdateStatus(lbmanagement.Stopping, "Stopping")
 			return
 		case change := <-cm.client.UnitChanges():
+			cm.logger.Debugf(
```
Comment (cmacknz) [inline x-pack/libbeat/management/managerV2.go:401]: I am open to doing this at the Info level as a type of audit log for when changes happened. If we only log actual changes, it should occur that frequently.

I'm not sure we need the `[features]` prefix either since this doesn't seem specific to the feature flags.

File: x-pack/libbeat/management/managerV2.go [hunk 20]
```diff
@@ -613,7 +636,7 @@ func (cm *BeatV2Manager) reloadInputs(inputUnits []*client.Unit) error {
 }
 
 // this function is registered as a debug hook
-// it prints the last known configuration genreated by the beat
+// it prints the last known configuration generated by the beat
```
Comment (cmacknz) [inline x-pack/libbeat/management/managerV2.go:639]: I think you will want to save the most recently received feature configuration and dump it out in this function. It can follow the same pattern the output configuration uses of saving it and then writing it out.

This function generates a file called `beat-rendered-config.yml` in the components sub-directories for Beats in the agent diagnostics, so you can inspect the Beat configuration that was generated based on the agent policy configuration. This is really valuable for debugging so we'll want to keep it up to date.

You will probably want to add a Features entry to the file structure. For an example of what this file looks like run `elastic-agent diagnostics`, unpack the archive, and search for the `beat-rendered-config.yml`. I can help you interpret the diagnostics if it isn't obvious what the folder structure is, the top level is documented in https://www.elastic.co/guide/en/fleet/current/elastic-agent-cmd-options.html#elastic-agent-diagnostics-command

https://github.com/elastic/beats/blob/e7e6dacfca2055d14fba269253dade2d197906d8/x-pack/libbeat/management/managerV2.go#L642-L650

**Review by cmacknz — COMMENTED** [review]
Looks good, a few minor things that can be improved.

It would still be great to get the agent management unit tests updated to include Feature processing: https://github.com/elastic/beats/blob/e7e6dacfca2055d14fba269253dade2d197906d8/x-pack/libbeat/management/managerV2_test.go#L72

This should save us (and the next developer in here) from having to manually test with the agent as often when making changes.

**Comment by ycombinator** [comment]
> It would still be great to get the agent management unit tests updated to include Feature processing:

I've been fiddling with https://github.com/elastic/beats/blob/e7e6dacfca2055d14fba269253dade2d197906d8/x-pack/libbeat/management/managerV2_test.go#L26, trying to incorporate feature flags into it.  Sure enough, it looks like we're missing something in the `elastic-agent-client` code - I don't think the observed checkin message contains features, which (IIUC) means the Agent (server) will keep re-sending features as part of the expected message to clients (components).  So I'm going to work on a PR to `elastic-agent-client` now in conjunction with amending the test in this PR to work with feature flags.

File: winlogbeat/sys/wineventlog/renderer.go (hunk 1 above)
Comment (ycombinator) [reply winlogbeat/sys/wineventlog/renderer.go:225]: Good catch, we don't need this TODO any more.  Removing...

File: x-pack/libbeat/management/managerV2.go (hunk 20 above)
Comment (ycombinator) [reply x-pack/libbeat/management/managerV2.go:639]: This is really helpful context, thank you, @cmacknz! I will make sure `beat-rendered-config.yml` contains feature flag information.

**Comment by ycombinator** [comment]
> ... Sure enough, it looks like we're missing something in the `elastic-agent-client` code - I don't think the observed checkin message contains features, which (IIUC) means the Agent (server) will keep re-sending features as part of the expected message to clients (components). So I'm going to work on a PR to `elastic-agent-client` now in conjunction with amending the test in this PR to work with feature flags.

Here's the PR to `elastic-agent-client`: https://github.com/elastic/elastic-agent-client/pull/63

And I've updated `TestManagerV2` to account for feature flags in this PR (see d193619cf62ab29f3361963de5dbcf371be24a6e).  Note that that commit currently contains a `replace` directive in `go.mod` to my personal fork of the `elastic-agent-client` repo.  I'll remove this directive once https://github.com/elastic/elastic-agent-client/pull/63 is reviewed and merged.

File: x-pack/libbeat/management/managerV2.go (hunk 20 above)
Comment (ycombinator) [reply x-pack/libbeat/management/managerV2.go:649]: This is implemented now.

Screenshot 2023-03-16 at 07 13 05

**Comment by ycombinator** [comment]
CI is failing on tests unrelated to this PR.  The same tests are currently failing in `main` as well.  They appear to be all related to the X-Pack Metricbeat `stan` module.  I've created https://github.com/elastic/beats/issues/34844 with more details about the failing test for the appropriate team to fix it.

File: x-pack/libbeat/management/managerV2.go [hunk 21]
```diff
@@ -443,24 +460,40 @@ func (cm *BeatV2Manager) reload(units map[unitKey]*client.Unit) {
...
-		if ll > lowestLevel {
+		expected := unit.Expected()
+		if expected.LogLevel > lowestLevel {
 			// log level is still used from an expected stopped unit until
 			// the unit is completely removed (aka. fully stopped)
-			lowestLevel = ll
+			lowestLevel = expected.LogLevel
 		}
-		if state == client.UnitStateStopped {
+		if expected.Features != nil {
```
Comment (cmacknz) [inline x-pack/libbeat/management/managerV2.go:472]: Do we disable the feature if `expected.Features == nil` and the current Features configuration is not nil? Is this a case you need to handle, or does the agent always provided a non-nil Features in expected?

Comment (ycombinator) [reply x-pack/libbeat/management/managerV2.go:472]: In the beginning, before any feature flags are set, Agent will send a `nil` value for `expected.Features`.  So this check here guards against that condition, preventing any unnecessary processing.  Thereafter, whenever a feature flag gets set, Agent will only send non-nil values for `expected.Features`; see https://github.com/elastic/elastic-agent/pull/2218/files#diff-5570d8d3d94d053efb7e59bcad53e531bcf3a0ff5b8d3da9cc3334d859f095c9R176-R181

Since `expected.Features` is built to contain multiple features' configuration in it, we would only disable a feature if that particular feature's configuration within `expected.Features` indicated as such.  For example, if `expected.Features.FQDN.Enabled` was set to `false`.  AFAIU we have not designed for a use case of mass-disabling all features by setting `expecting.Features` to `nil`.

Comment (cmacknz) [reply x-pack/libbeat/management/managerV2.go:472]: > AFAIU we have not designed for a use case of mass-disabling all features by setting expecting.Features to nil.

That doesn't mean someone won't make that edit in the policy. IMO the most obvious thing to do is to translate that to disabling all features, that maps to what happens when you remove an input from the policy. We can file a follow up issue to deal with this, it is definitely an edge case.
